File: apps/node-api/src/tasks.ts
/**
 * Analysis tasks executed by the job queue. Each manages its own job-state
 * updates and runs the shared engine over the ingested files. Mirrors the
 * Python API's tasks module.
 */

import { Engine, analyzeFiles } from "@aicodecheck/detector-core";

import { getStore } from "./db.js";
import {
  cloneRepository,
  extractZip,
  rmrf,
  walkDirectory,
} from "./ingestion.js";
import { getSettings } from "./config.js";
import { decryptToken } from "./security.js";

const engine = new Engine();

export interface RepositoryTaskPayload {
  jobId: string;
  reference: string;
  url: string;
  tokenEncrypted: string | null;
  ref: string | null;
  maxFiles: number;
  cacheKey: string | null;
}

export interface ZipTaskPayload {
  jobId: string;
  zipPath: string;
  cacheKey: string | null;
}

export async function analyzeRepositoryTask(p: RepositoryTaskPayload): Promise<string> {
  const store = await getStore();
  let clonePath: string | null = null;
  try {
    await store.updateJob(p.jobId, { status: "running", progress: 0.1 });
    const token = p.tokenEncrypted ? decryptToken(p.tokenEncrypted) : null;
    clonePath = await cloneRepository(p.url, token, p.ref);
    await store.updateJob(p.jobId, { progress: 0.5 });
    const files = await walkDirectory(clonePath, p.maxFiles);
    await store.updateJob(p.jobId, { progress: 0.7 });
    const result = engine.analyze(
      loadFolderScan(files),
      p.jobId,
    );
    // Attach owner/repo to the target so history/repository lookup works.
    annotateRepoTarget(result, p.reference);
    await store.saveAnalysis(result, { job_id: p.jobId, cache_key: p.cacheKey });
    await store.updateJob(p.jobId, { status: "finished", progress: 1 });
    return p.jobId;
  } catch (err) {
    await store.updateJob(p.jobId, {
      status: "failed",
      error: String((err as Error)?.message ?? err).slice(0, 2000),
    });
    throw err;
  } finally {
    if (clonePath) await rmrf(clonePath);
  }
}

export async function analyzeZipTask(p: ZipTaskPayload): Promise<string> {
  const store = await getStore();
  try {
    await store.updateJob(p.jobId, { status: "running", progress: 0.1 });
    const files = await extractZip(p.zipPath);
    await store.updateJob(p.jobId, { progress: 0.6 });
    const result = analyzeFiles(files, engine);
    // Re-id with the job id so /analysis/:jobId resolves.
    const withId = { ...result, id: p.jobId };
    await store.saveAnalysis(withId, { job_id: p.jobId, cache_key: p.cacheKey });
    await store.updateJob(p.jobId, { status: "finished", progress: 1 });
    return p.jobId;
  } catch (err) {
    await store.updateJob(p.jobId, {
      status: "failed",
      error: String((err as Error)?.message ?? err).slice(0, 2000),
    });
    throw err;
  } finally {
    await rmrf(p.zipPath);
    // best-effort: clean stale work dir contents older than ttl is omitted.
    void getSettings();
  }
}

// --- helpers --------------------------------------------------------------- //

import { loadFiles } from "@aicodecheck/detector-core";
import type { AnalysisResult } from "@aicodecheck/detector-core";

function loadFolderScan(files: Record<string, string>) {
  return loadFiles(files);
}

function annotateRepoTarget(result: AnalysisResult, reference: string): void {
  // reference is owner/repo or a URL; extract owner/repo for history filtering.
  const cleaned = reference
    .replace(/^https?:\/\/[^/]+\//i, "")
    .replace(/^git@[^:]+:/i, "")
    .replace(/\.git$/i, "");
  const parts = cleaned.split("/").filter(Boolean);
  if (parts.length >= 2) {
    result.target.owner = parts[parts.length - 2]!;
    result.target.repo = parts[parts.length - 1]!;
    result.target.name = `${result.target.owner}/${result.target.repo}`;
  }
  result.target.kind = "repository";
}

File: apps/python-api/app/tasks.py
"""Analysis tasks executed by the job queue (RQ worker or thread pool).

These are module-level functions with serializable arguments so they can be
enqueued by RQ. They each manage their own DB session and update job state.
"""

from __future__ import annotations

import os
import traceback

from aiprojectdetector import Engine
from aiprojectdetector.ingestion.loader import IngestOptions, load_files, load_github, load_zip

from .config import get_settings
from .db import session_scope
from .storage import save_analysis, update_job


def _opts(config: dict) -> IngestOptions:
    return IngestOptions(
        ignore_patterns=config.get("ignore_patterns", []),
        include_hidden=config.get("include_hidden", True),
        include_lockfiles=config.get("include_lockfiles", False),
        analyze_git=config.get("analyze_git", True),
        max_files=config.get("max_files", 50_000),
    )


def analyze_repository_task(
    job_id: str,
    reference: str,
    token_encrypted: str | None,
    config: dict,
    cache_key: str | None,
) -> str:
    settings = get_settings()
    os.makedirs(settings.work_dir, exist_ok=True)
    session = session_scope()
    try:
        update_job(session, job_id, status="running", progress=0.1)
        token = None
        if token_encrypted:
            from .security import decrypt_token

            token = decrypt_token(token_encrypted)
        scan = load_github(reference, _opts(config), token=token, workdir=settings.work_dir)
        update_job(session, job_id, progress=0.6)
        result = Engine().analyze(scan, analysis_id=job_id)
        save_analysis(session, result, job_id=job_id, cache_key=cache_key)
        update_job(session, job_id, status="finished", progress=1.0)
        return job_id
    except Exception as exc:  # noqa: BLE001
        update_job(session, job_id, status="failed", error=f"{exc}\n{traceback.format_exc()[:2000]}")
        raise
    finally:
        session.close()


def analyze_zip_task(job_id: str, zip_path: str, config: dict, cache_key: str | None) -> str:
    settings = get_settings()
    session = session_scope()
    try:
        update_job(session, job_id, status="running", progress=0.1)
        scan = load_zip(zip_path, _opts(config), workdir=settings.work_dir)
        update_job(session, job_id, progress=0.6)
        result = Engine().analyze(scan, analysis_id=job_id)
        save_analysis(session, result, job_id=job_id, cache_key=cache_key)
        update_job(session, job_id, status="finished", progress=1.0)
        return job_id
    except Exception as exc:  # noqa: BLE001
        update_job(session, job_id, status="failed", error=f"{exc}\n{traceback.format_exc()[:2000]}")
        raise
    finally:
        try:
            os.remove(zip_path)
        except OSError:
            pass
        session.close()


def analyze_files_task(job_id: str, files: dict, name: str, config: dict, cache_key: str | None) -> str:
    session = session_scope()
    try:
        update_job(session, job_id, status="running", progress=0.2)
        scan = load_files(files)
        scan.name = name
        result = Engine().analyze(scan, analysis_id=job_id)
        save_analysis(session, result, job_id=job_id, cache_key=cache_key)
        update_job(session, job_id, status="finished", progress=1.0)
        return job_id
    except Exception as exc:  # noqa: BLE001
        update_job(session, job_id, status="failed", error=str(exc))
        raise
    finally:
        session.close()
